File: cert-processor/src/main/java/org/sunbird/incredible/processor/signature/SignatureHelper.java
package org.sunbird.incredible.processor.signature;


import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.http.HttpHeaders;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.sunbird.incredible.processor.JsonKey;
import org.sunbird.incredible.processor.signature.exceptions.SignatureException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import javax.ws.rs.core.MediaType;


public class SignatureHelper {

    private final String SIGN_API_ENDPOINT;

    private final String VERIFY_API_ENDPOINT;

    private ObjectMapper mapper = new ObjectMapper();

    public SignatureHelper(String encServiceUrl) {
        SIGN_API_ENDPOINT = encServiceUrl.concat("/" + JsonKey.SIGN + "/");
        VERIFY_API_ENDPOINT = encServiceUrl.concat("/" + JsonKey.VERIFY);
    }

    private static Logger logger = LoggerFactory.getLogger(SignatureHelper.class);


    /**
     * This method calls signature service for signing the object
     *
     * @param rootNode - contains input need to be signed
     * @return - signed data with key
     * @throws SignatureException.UnreachableException
     * @throws SignatureException.CreationException
     */
    public Map<String, Object> generateSignature(JsonNode rootNode, String keyId)
            throws SignatureException.UnreachableException, SignatureException.CreationException {
        Map signReq = new HashMap<String, Object>();
        signReq.put(JsonKey.ENTITY, rootNode);
        CloseableHttpClient client = HttpClients.createDefault();
        logger.info("generateSignature:keyID:".concat(keyId));
        String encServiceUrl = SIGN_API_ENDPOINT.concat(keyId);
        logger.info("generateSignature:enc service url formed:".concat(encServiceUrl));
        HttpPost httpPost = new HttpPost(encServiceUrl);
        try {
            StringEntity entity = new StringEntity(mapper.writeValueAsString(signReq));
            logger.info("generateSignature:SignRequest for enc-service call:".concat(mapper.writeValueAsString(signReq)));
            httpPost.setEntity(entity);
            httpPost.setHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
            CloseableHttpResponse response = client.execute(httpPost);
            return mapper.readValue(response.getEntity().getContent(),
                    new TypeReference<Map<String, Object>>() {
                    });
        } catch (ClientProtocolException e) {
            logger.error("ClientProtocolException when signing: {}", e.getMessage());
            throw new SignatureException().new UnreachableException(e.getMessage());
        } catch (IOException e) {
            logger.error("RestClientException when signing: {}", e.getMessage());
            throw new SignatureException().new CreationException(e.getMessage());

        }

    }


    public boolean verifySignature(JsonNode rootNode)
            throws SignatureException.UnreachableException, SignatureException.VerificationException {
        logger.debug("verify method starts with value {}", rootNode);
        Map signReq = new HashMap<String, Object>();
        signReq.put(JsonKey.ENTITY, rootNode);
        boolean result = false;
        CloseableHttpClient client = HttpClients.createDefault();
        HttpPost httpPost = new HttpPost(VERIFY_API_ENDPOINT);
        httpPost.setHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
        try {
            StringEntity entity = new StringEntity(mapper.writeValueAsString(signReq));
            httpPost.setEntity(entity);
            CloseableHttpResponse response = client.execute(httpPost);
            result = mapper.readValue(response.getEntity().getContent(),
                    new TypeReference<Boolean>() {
                    });

        } catch (ClientProtocolException ex) {
            logger.error("ClientProtocolException when verifying: {}", ex.getMessage());
            throw new SignatureException().new UnreachableException(ex.getMessage());
        } catch (Exception e) {
            logger.error("Exception occurred  while verifying signature:{} ", e.getMessage());
            throw new SignatureException().new VerificationException("");
        }
        logger.debug("verify method ends with value {}", result);
        return result;
    }

}

File: certificate-processor/src/main/scala/org/sunbird/incredible/processor/signature/SignatureHelper.scala
package org.sunbird.incredible.processor.signature

import java.io.IOException

import com.fasterxml.jackson.core.`type`.TypeReference
import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import org.apache.http.client.ClientProtocolException
import org.slf4j.{Logger, LoggerFactory}
import org.sunbird.incredible.{HTTPResponse, HttpUtil, JsonKeys}


object SignatureHelper {

  var httpUtil = new HttpUtil
  private val logger: Logger = LoggerFactory.getLogger(getClass.getName)
  lazy private val mapper: ObjectMapper = new ObjectMapper()

  /**
    * This method calls signature service for signing the object
    *
    * @param rootNode - contains input need to be signed
    * @return - signed data with key
    * @throws SignatureException.UnreachableException
    * @throws SignatureException.CreationException
    */

  def generateSignature(rootNode: JsonNode, keyId: String)(implicit encServiceUrl: String): java.util.Map[String, AnyRef] = {
    val signReq: Map[String, AnyRef] = Map(JsonKeys.ENTITY -> rootNode)
    logger.info("generateSignature:keyID:".concat(keyId))
    val signApiURl: String = encServiceUrl.concat("/" + JsonKeys.SIGN + "/").concat(keyId)
    logger.info("generateSignature:enc service url formed:".concat(signApiURl))
    try {
      logger.info("generateSignature:SignRequest for enc-service call:".concat(mapper.writeValueAsString(signReq)))
      val response = httpUtil.post(signApiURl, mapper.writeValueAsString(signReq))
      mapper.readValue(response.body, new TypeReference[java.util.Map[String, AnyRef]]() {})
    } catch {
      case e: ClientProtocolException =>
        logger.error("ClientProtocolException when signing: {}", e.getMessage)
        throw new SignatureException.UnreachableException(e.getMessage)
      case e: IOException =>
        logger.error("RestClientException when signing: {}", e.getMessage)
        throw new SignatureException.CreationException(e.getMessage)
    }
  }


  def verifySignature(rootNode: JsonNode)(implicit encServiceUrl: String): Boolean = {
    val verifyEndPoint = encServiceUrl.concat("/" + JsonKeys.VERIFY)
    logger.debug("verify method starts with value {}", rootNode)
    val signReq: Map[String, AnyRef] = Map(JsonKeys.ENTITY -> rootNode)
    var result = false
    try {
      val response: HTTPResponse = httpUtil.post(verifyEndPoint, mapper.writeValueAsString(signReq))
      result = mapper.readValue(response.body, new TypeReference[Boolean]() {})
    } catch {
      case ex: ClientProtocolException =>
        logger.error("ClientProtocolException when verifying: {}", ex.getMessage)
        throw new SignatureException.UnreachableException(ex.getMessage)
      case e: Exception =>
        logger.error("Exception occurred  while verifying signature:{} ", e.getMessage)
        throw new SignatureException.VerificationException("")
    }
    result
  }

}
